From a1022442dc2f59321976fd3456ad97c61300d1f2 Mon Sep 17 00:00:00 2001
From: Ahmad Samir <a.samirh78@gmail.com>
Date: Sun, 14 Jun 2020 11:40:21 +0200
Subject: [PATCH] Don't add extra newlines when pasting from GTK applications

Due to some internal handling in GTK itself, we end up with text copied
from Firefox where a newline is represented as CRLF (\r\n); since the
doPaste() method replaces each \n with \r, we ended up with \r\r, which
ultimately led to extra newlines in the pasted code. To fix the issue
simply detect the CRLF case and replace each \r\n with \r.

To test, copy some multiline text from Firefox, open vim in konsole and
paste, you end up with two newlines between each block of text. Compare the
results after applying this patch

BUG: 421480
FIXED-IN: 20.08
---
 src/TerminalDisplay.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/TerminalDisplay.cpp b/src/TerminalDisplay.cpp
index cbb6f500..387061fb 100644
--- a/src/TerminalDisplay.cpp
+++ b/src/TerminalDisplay.cpp
@@ -3274,6 +3274,10 @@ void TerminalDisplay::doPaste(QString text, bool appendReturn)
     }
 
     if (!text.isEmpty()) {
+        // replace CRLF with CR first, fixes issues with pasting multiline
+        // text from gtk apps (e.g. Firefox), bug 421480
+        text.replace(QLatin1String("\r\n"), QLatin1String("\r"));
+
         text.replace(QLatin1Char('\n'), QLatin1Char('\r'));
         if (bracketedPasteMode()) {
             text.remove(QLatin1String("\033"));
-- 
2.27.0

